File: lab2/ex1/mymax.c
//
// Created by m43 on 22. 04. 2020..
//
#include <stdio.h>
#include <string.h>

int gt_int(const void *i1, const void *i2) {
    return *(int *) i1 > *(int *) i2 ? 1 : 0;
}

int gt_char(const void *c1, const void *c2) {
    return *(char *) c1 > *(char *) c2 ? 1 : 0;
}

int gt_str(const void *s1, const void *s2) {
    return strcmp(*(char **) s1, *(char **) s2) > 0 ? 1 : 0;
}

const void *mymax(
        const void *base, size_t nmemb, size_t size,
        int (*compar)(const void *, const void *)) {

    int index_of_max = 0;
    for (int i = 1; i < nmemb; i++) {
        if (compar(&base[index_of_max * size], &base[i * size]) != 1) {
            index_of_max = i;
        }
    }

    return &base[index_of_max * size];
}

int main() {
    int arr_int[] = {1, 3, 5, 7, 4, 6, 9, 2, 0};
    char arr_char[] = "Suncana strana ulice";
    const char *arr_str[] = {
            "Gle", "malu", "vocku", "poslije", "kise",
            "Puna", "je", "kapi", "pa", "ih", "njise"
    };

    int *max_int = mymax(arr_int, sizeof(arr_int) / sizeof(*arr_int), sizeof(int), gt_int);
    char *max_char = mymax(arr_char, sizeof(arr_char) / sizeof(*arr_char), sizeof(char), gt_char);
    char **max_str = mymax(arr_str, sizeof(arr_str) / sizeof(*arr_str), sizeof(char *), gt_str);

    printf("The max element of arr_int is: %d\n", *max_int);
    printf("The max element of arr_char is: %c\n", *max_char);
    printf("The max element of arr_str is: %s\n", *max_str);

    return 0;
}

File: lab2/ex3/mymax.py
def mymax(iterable, key=lambda x: x):
    max_x = max_key = None

    for x in iterable:
        current_key = key(x)
        if max_key is None or current_key > max_key:
            max_x = x
            max_key = current_key

    return max_x


if __name__ == "__main__":
    maxint = mymax([1, 3, 5, 7, 4, 6, 9, 2, 0])
    print("The max integer among the integers is:", maxint)
    maxchar = mymax("Suncana strana ulice")
    print("The biggest char among the chars:", maxchar)

    maxstring = mymax([
        "Gle", "malu", "vocku", "poslije", "kise",
        "Puna", "je", "kapi", "pa", "ih", "njise"])
    print("The biggest string among the strings:", maxstring)

    D = {'burek': 8, 'buhtla': 5}
    print("The most expensive product in the bakery on the corner is:",
          mymax(D, D.get))

    maxperson = mymax(
        [("Josip", "Petrović"), ("Will", "Smith"), ("Julián", "Seco"), ("Angela", "Luque"), ("Mauricio", "Dengra"),
         ("Jose", "Fraga"), ("Marco", "Venegas"), ("Marcos", "Arnal"), ("Iván", "Japón"), ("Gerardo", "Moruga"),
         ("Enrique", "Barrueco"), ("Alfredo", "Muñoz"), ("Bárbara", "Palacio")])
    print("Who is a legend? -->", maxperson[0], maxperson[1])

"""
    "Find the most expensive product in Dictionary D containing
    the price list of the bakery on the corner of the street:
        D = {'burek': 8, 'buhtla': 5}
    To accomplish this task, pass the get dictionary method to D
    as a key. Explain how and why we can use the method as a
    free function."
    Functions are objects in Python that can be freely passed
    around. Python is very flexible in regards to functions.
    For example, one can change the definition of functions in
    runtime or even define entirely new functions in runtime
    and bind them to a class.
"""
